CARDS-2441: Computed questions only have access to the first value of multivalued variables

marta- opened the pull request:
**Summary of changes:**
* introduced new syntax for returning the values of a variable as an array via `[]`: `@{[questionId:-defaultvalue]}` and for accepting missing values via `?`:  `@{[questionId?]}`
  * updated the `ComputedQuestion` **frontend code** to put the values in an array when the `[]` are used (even if the variable is single valued) and to accept missing values when `?` is used; when the already existing syntax is used ( `@{questionId:-defaultvalue}`, no square brackets, no questiob mark), the behavior remains the same, and only the first value is return from multivalued answers with the expression not being evaluated when any value is missing
* updated the `BackendComputedTest` questionnaire to showcase computed answers with multivalued variables, including with the `?` syntax
* updated the `ComputedTest` questionnaire to showcase one computed question using a `text` multivalued variable and one computed question using a `long` multivalued variable 

**Testing:**
* start with `--test`
* test the entire `ComputedTest` questionnaire (this will cover new syntax as well as backward compatibility)
* test the questionnaire `BackendComputedTest` following the instructions listed in the infobox as both `admin` and `computedtestuser`
  * **Question 4** and **Question 4a|b|c Computed** help test the newly introduced syntax, but test everything for backwards compatibility 
* testing as a developer: create your own computed question using multivalued variables
  * more interesting additions to the `ComputedTest` questionnaires are welcome

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables

Frontend: if more than one value is available, added all values as an array instead of only the first one

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables

Added test question to the BackendComputedTest questionnaire

sashaandjic commented:
`computedtestuser` isn't a trusted one:
![image](https://github.com/data-team-uhn/cards/assets/23243244/0b84a555-597b-4ff9-b326-f1281f5e8614)

once I moved the `computedtestuser` to the trusted group, I was not able to see the form that admin created:

![image](https://github.com/data-team-uhn/cards/assets/23243244/b24d2618-be8b-4812-8f61-2ad2bd5b6be3)

sashaandjic reviewed (approved):
tested
![image](https://github.com/data-team-uhn/cards/assets/23243244/bc609a39-0f2e-43fb-a01e-2afdeb9f9278)

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables

Introduced new syntax for getting values as an array: `@{[varname:-defaultvalue]}`

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables

Updated test questionnaires to cover newly introduced syntax for retrieving values as an array:
* BackendComputedTest: updated the expression of Question 4 Computed
* ComputedTest: added 2 multivalued questions (one text, one long) and two computed questions using those values

marta- commented:
The frontend code and testing instructions have been updated, please re-test @sashaandjic

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables

Expanded the syntax to add support for accepting missing values via `!`: `@{[!varname]}`

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables

Updated the syntax for accepting missing values via `?`: `@{varname?}` or `@{[varname?]}`

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -176,13 +197,21 @@ let ComputedQuestion = (props) => {
+      let acceptMissingValue = false;
+      if (questionName.endsWith(acceptMissingTag)) {
+        acceptMissingValue = true;
+        questionName = questionName.replace(acceptMissingTag, "");
```
Hopefully there's no question mark inside the question name itself.

sdumitriu reviewed (commented):
Looks good so far.

acrowthe committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Add support for backend handling of multivalued computed questions

acrowthe committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Clean up backend computed code

marta- committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Minor frontend refactoring

sashaandjic commented:
![image](https://github.com/data-team-uhn/cards/assets/23243244/3aa306c5-92d0-44c5-a79a-74240f48ad19)

sashaandjic commented:
Is this PR complete now?

marta- commented:
> Is this PR complete now?

Ready for review and testing.

sashaandjic commented:
Why is `Question 4b` only visible in `edit` mode?
![image](https://github.com/data-team-uhn/cards/assets/23243244/09464d24-f1dc-404a-814d-43610584ede7)

marta- commented:
> Why is `Question 4b` only visible in `edit` mode?

Javascript error in the backend. The frontend returns empty array when `?` is used and the value is missing, while the backend returns `null`. We need to decide which one we want to keep. I would say `[]` because the `?` is inside the brackets @acrowthe .

acrowthe committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
- Fix multivalued backend passing null rather than an empty array for optional arrays

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ExpressionUtilsImpl.java:
```diff
@@ -93,56 +93,138 @@ public Object evaluate(final Node question, final Map<String, Object> values, fi
+            this.startMarker = this.isArrayArgument ? START_MARKER_ARRAY : START_MARKER_SINGLE;
+            this.endMarker = this.isArrayArgument ? END_MARKER_ARRAY : END_MARKER_SINGLE;
+            this.start = this.expression.indexOf(this.startMarker);
+            this.end = this.expression.indexOf(this.endMarker);
```
It is possible to have a `}` somewhere in the expression before the start marker, since that is a very common syntax element.

```suggestion
            this.end = this.expression.indexOf(this.endMarker, this.start);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -135,34 +138,52 @@ let ComputedQuestion = (props) => {
-    while(start > -1 && end > -1) {
+    while (expr) {
+      start = expr.indexOf(startTagArrayVal);
+      if (start >= 0) {
```
Bug: `return @{single} + @{[array]}` will not parse the `single` argument since it prioritizes looking for arrays. The Java code does this right, it checks which one comes next.

acrowthe committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Fix issue where front end computed questions would skip past any single valued
  arguments if a multi valued argument was after it

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ExpressionUtilsImpl.java:
```diff
@@ -93,56 +93,138 @@ public Object evaluate(final Node question, final Map<String, Object> values, fi
+
+        if (value != null) {
+            if (value.getClass().isArray() && !shouldBeArray) {
+                value = ((Object[]) value)[0];
```
There's a potential `ArrayIndexOutOfBoundsException` lurking here. Add a check for the length of value.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ExpressionUtilsImpl.java:
```diff
@@ -93,56 +93,138 @@ public Object evaluate(final Node question, final Map<String, Object> values, fi
+                if (questionValue == null) {
+                    if (!isOptional) {
+                        this.missingValue = true;
+                    } else if (isOptional && this.isArrayArgument) {
```
We already know it is optional:
```suggestion
                    } else if (this.isArrayArgument) {
```

acrowthe committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Address code review comments

sdumitriu committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Potential bugfix when an answer is stored as an empty array

sdumitriu committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Frontend regression: empty answer is no longer considered missing

sdumitriu committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
Yet another missing answer bugfix

acrowthe committed:
CARDS-2441: Computed questions only have access to the first value of multivalued variables
- Fix backend computed question arrays not being the proper Javascript array type

sashaandjic commented:
re-tested

![image](https://github.com/data-team-uhn/cards/assets/23243244/1f2c1960-f491-43fb-833f-da31fff1b275)
